# 03 EDA — Products

Read-only inspection of `olist_products_dataset.csv`. No cleaning, charts, or feature engineering.

## 1. Import libraries

Import pandas for tabular inspection.

In [1]:
import pandas as pd

**Observation:** pandas is ready.

## 2. Load CSV

Load only this CSV into memory without changing the source.

In [2]:
file_path = '../data/raw/olist/olist_products_dataset.csv'
df = pd.read_csv(file_path)

**Observation:** The table is loaded; the raw CSV is unchanged.

## 3. Shape

Check table size.

In [3]:
df.shape

(32951, 9)

**Observation:** This reports the table's row and column count.

## 4. First 5 rows

Preview the raw structure.

In [4]:
df.head()

                         product_id  ... product_width_cm
0  1e9e8ef04dbcff4541ed26657ea517e5  ...             14.0
1  3aa071139cb16b67ca9e5dea641aaa2f  ...             20.0
2  96bd76ec8810374ed1b65e291975717f  ...             15.0
3  cef67bcfe19066a932b7673e239eb23d  ...             26.0
4  9dc1a7de274444849c219cff195d0b71  ...             13.0

[5 rows x 9 columns]

**Observation:** The preview confirms the row grain and field layout.

## 5. Column names

List available fields.

In [5]:
df.columns.tolist()

['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

**Observation:** These are the fields available for this table's analysis.

## 6. Data types

Review pandas data types.

In [6]:
df.dtypes

product_id                        str
product_category_name             str
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

**Observation:** Types show which fields are text, numeric, or converted datetime values.

## 7. Missing values

Count nulls by field.

In [7]:
df.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

**Observation:** Missing counts identify incomplete fields without changing them.

## 8. Missing percentage

Measure missingness relative to table size.

In [8]:
(df.isna().mean().mul(100).round(2)).sort_values(ascending=False)

product_category_name         1.85
product_description_lenght    1.85
product_name_lenght           1.85
product_photos_qty            1.85
product_weight_g              0.01
product_height_cm             0.01
product_length_cm             0.01
product_width_cm              0.01
product_id                    0.00
dtype: float64

**Observation:** Percentages show the relative scale of incomplete fields.

## 9. Duplicate rows

Count complete row duplicates.

In [9]:
df.duplicated().sum()

np.int64(0)

**Observation:** This is the number of repeated full rows; none are removed.

## 10. Duplicate primary keys

Validate the documented candidate key.

In [10]:
primary_key = ['product_id']
df.duplicated(subset=primary_key, keep=False).sum()

np.int64(0)

**Observation:** This counts rows involved in duplicate candidate keys for product_id.

## 11. Unique values

Inspect meaningful category or entity counts.

In [11]:
product_counts = pd.Series({'products': df['product_id'].nunique(), 'nonblank_categories': df['product_category_name'].nunique()})
print(product_counts)
df['product_category_name'].value_counts(dropna=False).head(10)

products               32951
nonblank_categories       73
dtype: int64


product_category_name
cama_mesa_banho           3029
esporte_lazer             2867
moveis_decoracao          2657
beleza_saude              2444
utilidades_domesticas     2335
automotivo                1900
informatica_acessorios    1639
brinquedos                1411
relogios_presentes        1329
telefonia                 1134
Name: count, dtype: int64

**Observation:** This gives a concise view of important distinct values.

## 12. Basic describe()

Generate a basic summary for all fields.

In [12]:
df.describe(include='all')

                              product_id  ... product_width_cm
count                              32951  ...     32949.000000
unique                             32951  ...              NaN
top     1e9e8ef04dbcff4541ed26657ea517e5  ...              NaN
freq                                   1  ...              NaN
mean                                 NaN  ...        23.196728
std                                  NaN  ...        12.079047
min                                  NaN  ...         6.000000
25%                                  NaN  ...        15.000000
50%                                  NaN  ...        20.000000
75%                                  NaN  ...        30.000000
max                                  NaN  ...       118.000000

[11 rows x 9 columns]

**Observation:** The summary describes completeness, distinctness, common values, and numeric ranges where applicable.

## 13. Table-specific observations

Each row describes one product. Categories, photos, weight, and dimensions can later support product-performance and logistics analysis.

## 14. Data quality observations

The primary key is valid and no exact rows repeat. There are 610 missing values in four product metadata fields and two missing values in each physical-measure field; 13 nonblank category values lack a translation match.